propagation/http_trace_context_propagator.go

```diff
@@ -0,0 +1,195 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package propagation
+
+import (
+	"encoding/hex"
+	"fmt"
+	"net/http"
+	"net/textproto"
+	"strconv"
+	"strings"
+
+	"go.opentelemetry.io/api/core"
+	apipropagation "go.opentelemetry.io/api/propagation"
+	"go.opentelemetry.io/api/tag"
+)
+
+const (
+	supportedVersion  = 0
+	maxVersion        = 254
+	traceparentHeader = "traceparent"
+)
+
+type httpTraceContextPropagator struct{}
+
+var _ apipropagation.TextFormatPropagator = httpTraceContextPropagator{}
+
+// CarrierExtractor implements CarrierExtractor method of TextFormatPropagator interface.
+//
+// It creates CarrierExtractor object and binds carrier to the object. The carrier
+// is expected to be *http.Request. If the carrier type is not *http.Request
+// then an empty extractor. Extract method on empty extractor does nothing.
+func (hp httpTraceContextPropagator) CarrierExtractor(carrier interface{}) apipropagation.Extractor {
+	req, ok := carrier.(*http.Request)
+	if ok {
+		return traceContextExtractor{req: req}
+	}
+	return traceContextExtractor{}
+}
+
+// CarrierInjector implements CarrierInjector method of TextFormatPropagator interface.
+//
+// It creates CarrierInjector object and binds carrier to the object. The carrier
+// is expected to be of type *http.Request. If the carrier type is not *http.Request
+// then an empty injector is returned. Inject method on empty injector does nothing.
+func (hp httpTraceContextPropagator) CarrierInjector(carrier interface{}) apipropagation.Injector {
+	req, ok := carrier.(*http.Request)
+	if ok {
+		return traceContextInjector{req: req}
+	}
+	return traceContextInjector{}
+}
+
+// HttpTraceContextPropagator creates a new propagator that propagates SpanContext
+// in W3C TraceContext format.
+//
+// The propagator is then used to create CarrierInjector and CarrierExtractor associated with a
+// specific request. Injectors and Extractors respectively provides method to
+// inject and extract SpanContext into/from the http request. Inject method encodes
+// SpanContext into W3C TraceContext Header and injects the header in the request.
+// Extract method extracts the header and decodes SpanContext.
+func HttpTraceContextPropagator() httpTraceContextPropagator {
+	return httpTraceContextPropagator{}
+}
+
+type traceContextExtractor struct {
+	req *http.Request
+}
+
+var _ apipropagation.Extractor = traceContextExtractor{}
+
+// Extract implements Extract method of propagation.Extractor interface. It extracts
+// W3C TraceContext Header and decodes SpanContext from the Header.
+func (tce traceContextExtractor) Extract() (sc core.SpanContext, tm tag.Map) {
+	if tce.req == nil {
+		return core.EmptySpanContext(), tag.NewEmptyMap()
```

Comment:
That's a lot of `core.EmptySpanContext(), tag.NewEmptyMap()` repetition. Maybe add some empty function like `func noExtract() (sc core.SpanContext, m tag.Map) {return}` (or `func noExtract() (core.SpanContext, tag.Map) {return core.EmptySpanContext(), tag.NewEmptyMap()}`) and use that?